### ISSUE TITLE
Tune the evaluation weights: fit the 768 piece-square numbers instead of copying them, and give the taper two genuinely different tables per piece

### ISSUE THREAD

**Issue by UnrealPhantasy** [UnrealPhantasy/Blunderbuss#102]
## Context

`engine/src/evaluation.rs` holds about 450 hand-copied numbers — the material values, six
piece-square tables, the passed-pawn schedules and the mobility weights. They come from the
Chess Programming Wiki's "simplified evaluation function" and **have never been tuned**. They
were a reasonable place to start; they are not a reasonable place to stop.

The concrete defect is at the bottom of the file:

```rust
const PST_MIDDLEGAME: [[i32; 64]; 6] = [PAWN, KNIGHT, BISHOP, ROOK, QUEEN, KING_MIDDLEGAME];
const PST_ENDGAME:    [[i32; 64]; 6] = [PAWN, KNIGHT, BISHOP, ROOK, QUEEN, KING_ENDGAME];
```

**Five of the six tables are literally the same array**, and `value()` returns a single number
per piece with no phase argument. The 24-step taper the module documents at length therefore
moves *only the king*. A pawn on the sixth rank is worth exactly the same with both queens on
the board as it is in a pure pawn endgame; a rook is worth 500 in both. Every engine that has
measured this finds the two ends far apart — a pawn and a rook are worth clearly more in an
endgame, a knight clearly less.

This matters here in particular because of where the engine is weak. Measured against an
unrestricted Stockfish oracle on positions from its own games, the engine loses **24 cp per
move in the middlegame against 36 cp for a Stockfish restricted to ~1900** (it is ahead), and
**45 cp against 28 cp in the endgame** (it is behind). The whole net weakness sits in one
phase, and it is the phase the taper currently cannot express.

## Why a fit and not a search

The evaluation is **linear in these weights**:

```
eval = trunc( ( phase * SUM(mg_i) + (24 - phase) * SUM(eg_i) ) / 24 )
```

summed over the pieces present, with `+/-` by colour. So "are these weights good" is a convex
problem, not a hyperparameter search. A sparse logistic regression converges in minutes, and
SPSA / genetic algorithms / simulated annealing are all the wrong tool for it.

Material and the piece-square tables are **perfectly collinear**: `evaluate` adds `value(piece)`
to every square of a table, so only their sum is identifiable. The fit therefore has one
combined number per (piece, square, phase); splitting it back into a material constant plus a
table is a presentation choice made when the Rust is written, not a result.

## Goal

Replace the hand-copied constants with fitted ones, with two genuinely different tables per
piece so the taper does something, and publish what the change is worth on a held-out set and
against the oracle.

## Scope

- Six middlegame and six endgame piece-square tables, fitted (768 combined numbers).
- `PASSED_MIDDLEGAME` / `PASSED_ENDGAME` (the halving applied to a blockaded passer keeps the
  term linear — it just halves the coefficient).
- `MOBILITY_MIDDLEGAME` / `MOBILITY_ENDGAME` for the pieces that already carry a non-zero
  weight. Rook and queen mobility stay at zero: weighting them was measured to **double** the
  tree (733 605 nodes against 371 194 at depth 10 from the start position), and that decision
  is not reopened by a fit that cannot see node counts.
- The tests that pin the tables' shape, updated where a fitted table legitimately breaks an
  assertion written for the hand-made one — with the reason stated in the test.

### Out of scope

- **The search is not touched.** No change to reductions, pruning, ordering or the
  transposition table.
- `DRAWISH_DIVISOR` is not fitted. It is applied when building the features (a `cannot_mate`
  row has all its coefficients divided by 8, which stays linear as long as the divisor is not
  itself free), but the constant keeps its measured value.
- No new evaluation term. This issue makes the existing numbers right; it does not add
  knowledge.
- **No borrowed tables.** Not PeSTO, not Stockfish's. Stockfish is GPL-3.0 and contaminating —
  the same reason this project rejected shakmaty for the rules layer. And a table tuned for
  another engine encodes what *that* engine has no separate term for: PeSTO has no passed-pawn
  term and this engine does, so pasting its pawn table in would count the pawn race twice.

## Method

1. **Positions.** ~355 000 games of arena PGNs already on disk, 400 distinct book openings.
   **At most one position per game** and a cap per file — nine positions from one game share
   its pawn structure and its opponent, so they are not nine observations. The count of games
   is published next to the count of positions.
2. **Calm.** Each sampled position is replaced by the leaf of its own quiescence search, run
   through the engine's own `tactical_moves` and SEE filter, so "calm" means what the engine
   means by it. Leaves in check or terminal are dropped.
3. **Labels: a local Stockfish at depth 12, not the game result.** The result was produced by
   moves the current evaluation chose, so fitting to it would teach the weights their own
   preferences back. A Stockfish score is independent of them and graded rather than
   0 / 0.5 / 1. No Stockfish-vs-Stockfish games are generated: they are drowned in draws and
   their distribution is not the one this engine meets.
4. **Target in sigmoid space**, not raw centipawns: minimise the cross-entropy between
   `y = sigmoid(cp_sf / 400)` (with `cp_sf` clamped to +/-800) and `p = sigmoid(K * eval / 400)`.
   Squared error in centipawns over-weights positions that are already decided. `K` is fitted
   **separately and on both sides** of the comparison, then frozen while the weights move.
5. **20 % of the positions held out.** A loss drop measured on the training set is not a result.
6. **L2 regularisation toward the current values**, weak, strength chosen on the held-out set.

## The two unidentifiable directions, and how they are fixed

Both are directions along which the loss is exactly flat, so the optimiser drifts along them
on noise alone and the output stops being readable.

1. **The king gauge.** Every position has exactly one king of each colour, so adding a constant
   `C` to all 64 squares of the king table adds `+C` for White and `-C` for Black: the
   evaluation is *rigorously* unchanged. Left free, the fitted table reads "king = -23 on
   average" and invites an interpretation of what is an arbitrary point on a flat line. Fixed by
   subtracting the mean of the 64 values from each of them, middlegame and endgame separately,
   which makes the table read as "this square is worth +30 against the average square". The king
   is the only piece for which this is guaranteed — for the others the two sides do not always
   have the same number, so the direction is not flat.
2. **The scale gauge.** `p = sigmoid(K * eval / 400)`: multiplying every weight by `L` and
   dividing `K` by `L` gives exactly the same prediction. Fixed by fitting `K` first and
   freezing it, which is also what makes the before/after comparison fair.
3. **Frozen outright:** ranks 1 and 8 of the pawn table, 16 squares a pawn can never occupy.
   Their feature columns are identically zero, so anything written there is noise.

## The control that comes before any fitting

The feature construction must reproduce `evaluate()` **to the centipawn**. If it does not, the
fit optimises weights for a model the engine does not compute, the loss drops nicely and the
engine plays worse with nothing to say so.

This harness makes that check total rather than sampled: the feature extractor is written in
Rust next to `evaluate()`, and the consumer reconstructs the score from `sum_j w[j] * c[j]`
using the **same integer arithmetic** — taper truncating toward zero, drawish divisor
truncating again, side-to-move flip — and asserts equality on **100 % of the corpus**.

The five places a reimplementation diverges in practice, each reproduced explicitly:

- `Square::relative_to(color)` flips the **rank** for Black, not the file;
- the taper's division is integer and truncates toward zero, which differs from a floor for
  negative balances;
- `cannot_mate()` divides the balance by 8, and forgetting it makes every endgame row eight
  times too large;
- a **blocked** passed pawn has its bonus halved, on an `i32`, so an odd bonus loses half a
  centipawn — the blocked occurrences are counted in their own columns for that reason;
- `evaluate()` returns the **side-to-move** point of view, not White's.

## Acceptance criteria

- [x] `PST_MIDDLEGAME` and `PST_ENDGAME` no longer share a single array for five of six pieces:
      the twelve tables are independent, and `evaluate` is unchanged in structure.
- [x] The feature model reproduces `evaluate()` exactly on 100 % of the tuning corpus, and the
      figure is reported.
- [ ] The king tables are zero-mean, middlegame and endgame separately.
- [x] Ranks 1 and 8 of the pawn table are zero.
- [x] Held-out cross-entropy reported **before and after, each with its own optimal K**, on
      positions never seen by the fit, with the number of positions **and the number of games**.
- [x] The "after" figure is measured on scores produced by the **rebuilt engine**, not by the
      Python model — the float model is only an approximation of the integer evaluation.
- [ ] Centipawn loss against an unrestricted Stockfish oracle, per phase stratum, before and
      after, with its restricted-Stockfish control.
- [x] `cargo test` green. Every test whose assertion a fitted table breaks is either fixed or
      re-justified in place; none is deleted to make the suite pass.
- [ ] An anchored measurement against the calibrated anchor, and its confidence interval.

## Prediction, published before the first measurement

- Held-out cross-entropy: **-3 % to -6 %** relative.
- Oracle centipawn loss: the endgame stratum improves by **3 to 10 cp**, the middlegame stratum
  by less, possibly by nothing.
- **Anchored Elo: +25, interval [0, +55]** — the floor genuinely includes zero. Two reasons for
  the modest centre. A hand-made taper of these same tables was measured at **-1.8 +/- 1.3 cp**
  of oracle loss, i.e. nothing, so the direction alone is not enough. And an evaluation bias
  that is *monotone* changes no move ordering and is worth exactly 0 Elo — only **dispersion**
  costs, so part of any loss improvement will not convert.

**Stopping rule, fixed before measuring:** if the anchored interval over 3 200 games is
contained in [-10, +10], this brick closes and does not reopen without a new argument.

## Implementation notes

- The tuning harness (sampler, quiescence descent, labeller, fitter, Rust generator) lives
  **out of tree**, with a README carrying the figures and the traps hit. Only the generated
  constants and the tests land in this repository.
- Do not expect the node count to be preserved. Changing the evaluation moves where reverse
  futility pruning and forward futility pruning cut; a single table change was measured at
  **+15.9 % nodes** on one endgame position. The node count is not the control for this change.
- `cargo build --release` for every measurement. Debug figures are 20 to 50 times too slow.
